### PR TITLE
Add leakyrelu support

### PR DESCRIPTION
**Pull request by ljfitz**
Following the template of other recent additions.

### CODE REVIEW

File: e2e_testing/torchscript/elementwise.py [hunk 1]
```diff
@@ -4,6 +4,7 @@
 # Also available under a BSD-style license. See LICENSE.
 
 import torch
+import torch.nn.functional
```
Comment (pashu123) [inline e2e_testing/torchscript/elementwise.py:7]: Nit: This is not needed.

**Review by pashu123 — COMMENTED** [review]
Overall LGTM.

File: e2e_testing/torchscript/elementwise.py [hunk 2]
```diff
@@ -196,6 +197,24 @@ def ElementwiseReluModule_basic(module, tu: TestUtils):
...
+    def __init__(self):
+        super().__init__()
+
+    @export
+    @annotate_args([
+        None,
+        ([-1, -1], torch.float32, True),
+    ])
+    def forward(self, x):
+        return torch.nn.functional.leaky_relu(x, negative_slope=0.1)
```
Comment (cathyzhyi) [inline e2e_testing/torchscript/elementwise.py:210]: we could use `torch.ops.aten.leaky_relu` to test the op directly.

File: lib/Conversion/TorchToLinalg/TorchToLinalg.cpp [hunk 3]
```diff
@@ -1383,6 +1384,24 @@ static Value createLinalgPayloadCalculationForElementwiseOp(
...
+    if (!lrelu.getType()
+             .cast<ValueTensorType>()
+             .getDtype()
+             .isa<mlir::FloatType>()) {
+      lrelu.emitError("unimplemented: non-floating point dtype");
+      return nullptr;
+    }
+    Type elementType = payloadArgs[0].getType();
+    Value constZero =
+        b.create<arith::ConstantOp>(loc, FloatAttr::get(elementType, 0.0));
```
Comment (cathyzhyi) [inline lib/Conversion/TorchToLinalg/TorchToLinalg.cpp:1397]: nit: could use `getZeroAttr`

File: lib/Conversion/TorchToLinalg/TorchToLinalg.cpp [hunk 4]
```diff
@@ -1383,6 +1384,24 @@ static Value createLinalgPayloadCalculationForElementwiseOp(
...
+    Value pred = b.create<arith::CmpFOp>(loc, arith::CmpFPredicate::UGT,
+                                         payloadArgs[0], constZero);
+    Value positivePart = b.create<SelectOp>(loc, pred, payloadArgs[0], constZero);
+    Value negativePart = b.create<SelectOp>(loc, pred, constZero, payloadArgs[0]);
+    Value scaledNegativePart = b.create<arith::MulFOp>(loc, negativePart, payloadArgs[1]);
```
Comment (cathyzhyi) [inline lib/Conversion/TorchToLinalg/TorchToLinalg.cpp:1402]: The second arg is scalar and should be got from operands[1]. The `payloadArgs` are elements of tensor argument only.

File: python/torch_mlir/dialects/torch/importer/jit_ir/csrc/CMakeLists.txt [hunk 5]
```diff
@@ -30,6 +30,7 @@ target_link_libraries(TorchMLIRJITIRImporter
 
 message(STATUS "TORCH_CXXFLAGS=${TORCH_CXXFLAGS}")
 set_target_properties(TorchMLIRJITIRImporter PROPERTIES
+  INSTALL_PREFIX python_packages/torch_mlir/torch_mlir/_mlir_libs
```
Comment (cathyzhyi) [inline python/torch_mlir/dialects/torch/importer/jit_ir/csrc/CMakeLists.txt:33]: why is this needed here?

File: lib/Dialect/Torch/Transforms/RefineTypes.cpp [hunk 6]
```diff
@@ -224,7 +224,7 @@ class TypeAnalyzer : public ForwardDataFlowAnalysis<ValueKnowledge> {
   visitOperation(Operation *op,
                  ArrayRef<LatticeElement<ValueKnowledge> *> operands) final {
     if (isa<TensorStaticInfoCastOp, CopyToValueTensorOp, CopyToNonValueTensorOp,
-            AtenTanhOp, AtenBatchNormOp, AtenReluOp, AtenGeluOp,
+            AtenTanhOp, AtenBatchNormOp, AtenReluOp, AtenLeakyReluOp, AtenGeluOp,
```
Comment (cathyzhyi) [inline lib/Dialect/Torch/Transforms/RefineTypes.cpp:227]: The op involves both tensor and scalar so should use `visitBinaryTensorScalarOp` to take care of type promotion.

**Review by cathyzhyi — CHANGES_REQUESTED** [review]

**Comment by cathyzhyi** [comment]
> Following the template of other recent additions.
> 
> My test in `e2e_testing/torchscript/elementwise.py` is failing, and I'd love tips on how to debug.

I added some comments that might fix your failure.

In general, the Basic E2E debugging walk-through from https://github.com/llvm/torch-mlir/wiki covers most of the debugging tips. If you need help with debugging in the future, more details need to be provided like what is the error msg, what's the backtrace in python or c++ and IR snippets when applicable.

File: lib/Conversion/TorchToLinalg/TorchToLinalg.cpp (hunk 4 above)
Comment (ljfitz) [reply lib/Conversion/TorchToLinalg/TorchToLinalg.cpp]: Thanks.

File: python/torch_mlir/dialects/torch/importer/jit_ir/csrc/CMakeLists.txt (hunk 5 above)
Comment (ljfitz) [reply python/torch_mlir/dialects/torch/importer/jit_ir/csrc/CMakeLists.txt]: It wasn't meant to be in this PR :-( I've removed it.

**Comment by ljfitz** [comment]
> I added some comments that might fix your failure.

Thanks @cathyzhyi. The changes you suggested result in a pass. I had watched the debugging video but didn't understand how to apply it to understand the failure in `elementwise.py`. Luckily I don't need to for this PR - I'll rewatch for the next time.

**Comment by ljfitz** [comment]
I had opened this as a draft PR as I hadn't a working test at that moment. I wanted to remove the draft tag, but I don't see how. Could it be that I lack permissions?

**Comment by cathyzhyi** [comment]
Screen Shot 2021-11-24 at 5 08 08 PM

hmm, It's showing this one my side. Does it looks different for you?

**Comment by ljfitz** [comment]
> are you able to see the `ready for review`?

Yes, I see it now. Thanks again for pointing me to where I need to look!

**Review by cathyzhyi — APPROVED** [review]
LGTM. Thanks!

**Comment by cathyzhyi** [comment]
@ljfitz after the test passed and you are ready to merge, please @ me so that I can merge it for you.

**Comment by ljfitz** [comment]
@cathyzhyi the test passaat and I'm ready for this to merge.
